Update TF Decision Forests example

ksalama opened the pull request:
Hello @fchollet and @achoum - 

Please review the updates I made to the TF Decision Forests example.

Cheers

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -126,14 +115,39 @@
 """
 
-GROWING_STRATEGY = "BEST_FIRST_GLOBAL"
 NUM_TREES = 250
```
Please remove.
See comment in https://github.com/keras-team/keras-io/pull/782

ksalama replied:
I just wanted to expose some of the key parameters of the algorithms to be visible to the readers. I can add descriptions as comments to each one. Also happy to remove them. What do you think?

ksalama replied:
Do you think it is better to use the [documentation](https://www.tensorflow.org/decision_forests/api_docs/python/tfdf/keras/GradientBoostedTreesModel) link instead? (next to the model construction object)

achoum replied:
Yes it is equally good.

Note that this documentation pages also list all the Keras model arguments (not just the hyper-parameters) and it does not list the possible values the say way.

ksalama replied:
Done

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -100,6 +87,8 @@
+# The labels of the target columns.
+TARGET_LABELS = [" - 50000.", " 50000+."]
 # Weight column name.
 WEIGHT_COLUMN_NAME = "instance_weight"
```
Please remove.
See comment in https://github.com/keras-team/keras-io/pull/782

ksalama replied:
The `TARGET_LABELS` is used later in the `prepare_dataframe` method.

ksalama replied:
I am not using the attribute "as an input feature", rather as a weight to each instance. This will affect the loss computation; instances with higher weight will contribute more to the loss. This enforces the model to try to get the highly-weighted instances right, more than the low-weight instances, and I think it should be used as it represents the true distribution of the data.

Happy to discuss further

achoum replied:
I understood that.

However, do you think that `This attribute should *not* be used in the classifiers.` means that this field can still be used to do weighted training? This is effectively "using it" in the classifier.
I am not an expert in this dataset, so maybe is it that. Please confirm.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -146,23 +160,14 @@
-    return features, target, weight
-
-
 def run_experiment(model, train_data, test_data, num_epochs=1, batch_size=None):
```
What about implementing the "Nit: What about having a **fit_args argument to the run_experiment function?" discussed before.

ksalama replied:
Most of the Keras examples use the current `run_experiment` signature, and I didn't want to deviate from them.
Are you OK with that?

achoum replied:
I suspect that this is because those examples are training Neural networks (or another gradient descent based algorithm with mini-batching).

fchollet commented:
@achoum @ksalama are we good here?

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -146,23 +160,14 @@
+    )
 
     model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
     _, accuracy = model.evaluate(test_dataset, verbose=0)
```
At line 222 (for some reason, I cannot create a comment there):

You are effectively replicating the detection of the semantic of the features. This version is relatively simple and will not handle many cases. It is better to let TF-DF do the semantic detection.

achoum replied:
**Cannot comment in this block anymore.

>> Sort

I think tf.sort and unique are done in memory.
It would be interesting to understand and document what append in this processing layer when used on a large dataset. For example, will the entire dataset be loaded in memory?

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -146,23 +160,14 @@
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
 
     model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
```
This division will be unstable for rare categorical values. With target mean, it works well to add a correction term controlled by an hyperparameter.

correction =  1
positive_probability = positive_frequency / (positive_frequency + negative_frequency + correction)

ksalama replied:
Done

ksalama replied:
Done

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -146,23 +160,14 @@
-    ).map(prepare_sample, num_parallel_calls=tf.data.AUTOTUNE)
+    )
 
     model.fit(train_dataset, epochs=num_epochs, batch_size=batch_size)
```
In this example, I would do the division in double (float64) and then cast the result to float32.

ksalama replied:
Will do

ksalama replied:
Done

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -383,22 +387,22 @@ def call(self, inputs):
-            [positive_fequency, negative_fequency, positive_probability], axis=1
+            [positive_frequency, negative_frequency, positive_probability], axis=1
         )
 
```
Copy from https://github.com/keras-team/keras-io/pull/782

Could the colab report the test accuracy of the three tested methods?

Example: https://www.tensorflow.org/decision_forests/tutorials/model_composition_colab#whats_next

I don't see the output of the colab in this review.


>> TensorFlow Decision Forests provide powerful models, especially with structured data.
In our experiments, the Gradient Boosted Tree model achieved 95.79% test accuracy.
When using the target encoding with categorical feature, the same model achieved 95.81% test accuracy.
When pretraining embeddings to be used as inputs to the Gradient Boosted Tree model,
we achieved 95.82% test accuracy.

If the original model has a better accuracy than the model with pre-processing, this might indicate a bug in the code. I would not leave it this way. Could it be the smoothing hyper-parameter?

Alternatively, I would separate the simple GBT code from the experimental pre-processing in two separate examples (files). Users that just want to learn about GBT would benefit from it.

achoum commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -383,22 +387,22 @@ def call(self, inputs):
         # Concatenate and return the looked-up statistics.
         return tf.concat(
-            [positive_fequency, negative_fequency, positive_probability], axis=1
+            [positive_frequency, negative_frequency, positive_probability], axis=1
```
Can you address the comment starting with:

> This is not really an embedding. Instead, it will be the output of the linear model trained with categorical buckets. So it is fundamentally very similar to target mean.

in https://github.com/keras-team/keras-io/pull/782

ksalama replied:
OK I added a ReLU non-linearity to the encoder, and called the it NN model, rather than linear model

ksalama commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -146,23 +160,14 @@
-    return features, target, weight
-
-
 def run_experiment(model, train_data, test_data, num_epochs=1, batch_size=None):
```
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -126,43 +115,66 @@
+SAMPLING_METHOD = (
+    "RANDOM"  # Control the sampling of the datasets used to train individual trees.
+)
+VALIDATION_RATIO = 0.1  # Ratio of the training dataset used to monitor the training. Require to be >0 if early stopping is enabled.
```
In this section, move the comment descriptions to be above the variable (to keep lines short). Long descriptions should be split over 2 lines

ksalama replied:
Done

fchollet commented on examples/structured_data/classification_with_tfdf.py:
```diff
@@ -126,43 +115,66 @@
-                # Convert categorical feature values to string.
-                features[feature_name] = tf.strings.as_string(features[feature_name])
-    return features, target, weight
+The `run_experiment` method is responsible loading the train and test datasets,
```
`run_experiment()` (add `()`)

ksalama replied:
Done

fchollet reviewed (approved):
Thank you! LGTM. There are a couple style nits, once address you can add the generated files.

ksalama commented:
@achoum - Thank you so much for your comments. I addressed then as follows:

1. I used `CATEGORICAL_FEATURE_NAMES` instead of  `CATEGORICAL_FEATURES_WITH_VOCABULARY`
2. Create the feature semantics based on `NUMERICAL_FEATURE_NAMES` and `CATEGORICAL_FEATURES_WITH_VOCABULARY`, rather than the model input data types.
3. Deferred computing the vocabulary of the categorical features to the `create_embedding_encoder()` method.
4. Fixed the casting to float32 or float64 as we discussed
5. Updated the `BinaryTargetEncoding` to pre-compute all the required stats in the `adapt()` method, while the `call()` method would only do the lookup of the stats
6. Allow the user to provide `vocabulary_size` to avoid calling `tf.unique()`. However, if it is not passed, it's computed during the execution of the `adapt()` method. Also added a comment to recommend passing the `vocabulary_size`.
7. I kept the `prepare_dataframe()` method as it seems that ` tfdf.keras.pd_dataframe_to_tf_dataset` doesn't do all the "magic" of converting the target column to integer etc. Am I missing something?
8. Also, in the last model, when I perform activation directly on the feature, then linear projection, the results are worse than using `Dense(units, activation)(encoded_features)`. So I kept it as it was.
9. I haven't use `tf.data.experimental.group_by_reducer`, but I am planning to investigate it (in a subsequent PR).
10. Also, didn't use a dictionary for the parameters of the model as it, for some reasons, makes the example more than 300 lines, which is not allowed for the Keras examples!

Let me know if you have any further comments.

achoum commented:
All good :) Thanks for the work. This is awesome.

fchollet reviewed (approved):
Thanks a lot to both of you! 👍
